### PR TITLE
Implement default cache location

### PR DESCRIPTION
**Pull request by oscr**
Here is a draft for a possible solution for #234. 

I've added the big operating systems for now, but it can easily be expanded. I've ensure that if you enable cache yourself it isn't activated and if we don't recognize the platform it won't trigger.

On a machine without the cache folder existing it looks like this.

```
python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1
>>> session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscaru/.cache/fastf1.
	Size: 20.0 KB
```

After I've loaded a session. The next time I start fastf1 I will get the size. 

```
python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1
>>> session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscaru/.cache/fastf1.
	Size: 46.12 MB
```

If I explicitly enable caching then default caching won't active:

```
python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1
>>> fastf1.Cache.enable_cache("~/.fastf1")
>>> session = fastf1.get_session(2022, 13, 'Q')
>>> 
```

What do you think about this general approach?

### CODE REVIEW

**Comment by oscr** [comment]
Looks like test fail because I use `match` introduced in (Python 3.10)[https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching]. I can implement it on a backwards compatible way or if we upgrade Python version. Whatever works best for you.

**Comment by oscr** [comment]
Thinking about it: maybe it's better to not use newer features to avoid forcing users to upgrade.

**Comment by theOehrly** [comment]
> Thinking about it: maybe it's better to not use newer features to avoid forcing users to upgrade.

I haven't had time to look at this PR yet. But we can absolutely not require Python 3.10.

**Comment by oscr** [comment]
That won't be a problem. It's a trivial change from match to just if, elif etc.

**Comment by oscr** [comment]
When you have had a change to look at the general approach I will just change it. Less than 1 min change.

**Comment by oscr** [comment]
@theOehrly I noticed that some parts are not Python 3.10 compatible. Would it be possible to get a ticket outlining the issues with 3.10? I got 3.10 as default via Ubuntu LTS so even if it's not required, it's good if it works with that version. I could try to make some fixes if you have known issues.

**Comment by theOehrly** [comment]
> @theOehrly I noticed that some parts are not Python 3.10 compatible. Would it be possible to get a ticket outlining the issues with 3.10? I got 3.10 as default via Ubuntu LTS so even if it's not required, it's good if it works with that version. I could try to make some fixes if you have known issues.

Someone else just recently asked about that. I've explained the problem here: #245

**Comment by oscr** [comment]
Just to ensure we're not both waiting for each other: I'm waiting for a general ok of this approach before updating this pr.

**Comment by theOehrly** [comment]
I just still haven't gotten around to looking at this. But it's totally fine that you ask. Better than things waiting around for no reason.
Next week I have much more spare time, so I'll get back to you in a few days.

**Comment by theOehrly** [comment]
As a first draft this is fine. But there are a few things that need to be addressed.

1. As already stated, Py 3.8 and 3.9 compatibility
2. `get_size`, `convert_size` should be private and I'd prefer them to be staticmethods of `Cache`
3. I also think that it would be nice to add support for setting the cache path through an environment variable like `FASTF1_CACHE` or similar. This way, users can globally override the cache directory. The environment variable should have lower priority than a call to `Cache.enable_cache` but higher priority than the default location. I think this shouldn't be too difficult to add as environment variables should be platform independent I think.
4. %LOCALAPPDATA% wasn't expanded to the actual path when I just tried this on windows. If you don't have windows machine to test on, then I can investigate what's the problem there.

**Comment by oscr** [comment]
Thanks for the feedback! Should be no problem to fix this. 

Great find with nr 4! I don't have window but I found this: https://stackoverflow.com/questions/13184414/how-can-i-get-the-path-to-the-appdata-directory-in-python I will implement it and maybe you can verify it works when the pr is updated.

**Comment by oscr** [comment]
Changes have been implemented as discussed above. 

@theOehrly Could you please verify that the Windows fix actually works on Windows? I don't have any Windows machine.

When setting environment variable:

```
FASTF1_CACHE="~/ff1_env_cache" python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1
>>> session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscar/ff1_env_cache.
	Size: 20.0 KB
>>> 
```

Without environment variable
```
oscaru@Loke:~/repo/github/fast-f1$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1
>>> session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscar/.cache/fastf1.
	Size: 32.0 KB
>>> 
```

File: fastf1/api.py [hunk 1]
```diff
@@ -41,38 +43,37 @@
...
-
 base_url = 'https://livetiming.formula1.com'
 
 headers = {
-  'Host': 'livetiming.formula1.com',
-  'Connection': 'close',
-  'TE': 'identity',
-  'User-Agent': 'BestHTTP',
-  'Accept-Encoding': 'gzip, identity',
+    'Host': 'livetiming.formula1.com',
```
Comment (theOehrly) [inline fastf1/api.py:49]: You seem to have auto-formatted the whole file. While not wrong, I'd prefer to not have all these changes in this PR if possible.

**Comment by theOehrly** [comment]
%LOCALAPPDATA% now works correctly on Windows. 

The documentation of `Cache` needs to be updated as well, to reflect these changes.

Apart from that and my other review comments, everything looks good. Which operating systems have you tested this on?

File: fastf1/api.py (hunk 1 above)
Comment (oscr) [reply fastf1/api.py:49]: Looks like an IDEA bug to be honest. Didn't show up as changed in my editor but they where changed. Reverted.

File: fastf1/api.py [hunk 2]
```diff
@@ -318,16 +322,42 @@ def _write_cache(cls, data, cache_file_path, **kwargs):
...
+                    cache_dir = "~/Library/Caches/fastf1"
+                elif sys.platform == "win32":
+                    cache_dir = "%LOCALAPPDATA%/Temp/fastf1"
+
+            if cache_dir is not None:
+                # Ensure the default cache folder exists
+                cache_dir = os.path.expandvars(cache_dir)
+                cache_dir = os.path.expanduser(cache_dir)
+                if not os.path.exists(cache_dir):
+                    os.makedirs(cache_dir)
```
[the remark this answers is not in the document]
Comment (oscr) [reply fastf1/api.py:343]: Given the user can delete the .cache folder it's not guaranteed to exist.

If you'd prefer I would implement a Linux specific check: if ~/.cache exists use it, otherwise just place it in ~/.

[the remark this answers is not in the document]
Comment (oscr) [reply fastf1/api.py]: If `.cache` exists
```
python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1; session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscar/.cache/fastf1.
	Size: 32.0 KB
>>> 
```

If it doesn't
```
python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fastf1; session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscar/.fastf1.
	Size: 321.44 MB
>>> 
```

[the remark this answers is not in the document]
Comment (theOehrly) [reply fastf1/api.py]: I'm fine with that. It doesn't really make the logic too much more complicated.
Reagarding Windows, while one can in theory delete the temp folder, one really shouldn't. From a bit of quick googling, it sounds like FastF1 breaking might be the least of your concerns then.
The apple documentation phrases it like you can take it as a given that the folder exists. So that should be fine.

File: fastf1/api.py [hunk 3]
```diff
@@ -318,16 +322,47 @@ def _write_cache(cls, data, cache_file_path, **kwargs):
...
+                    cache_dir = r"~/Library/Caches/fastf1"
+                elif sys.platform == "win32":
+                    cache_dir = r"%LOCALAPPDATA%\Temp\fastf1"
+
+            if cache_dir is not None:
+                # Ensure the default cache folder exists
+                cache_dir = os.path.expandvars(cache_dir)
+                cache_dir = os.path.expanduser(cache_dir)
+                if not os.path.exists(cache_dir):
+                    os.mkdir(cache_dir, mode=0o0700)
```
Comment (theOehrly) [inline fastf1/api.py:348]: We should use some kind of improved error handling here, if the directory creation fails. I'd say just catch the error and re-raise it, but state specifically that the cache directory creation failed and state which path it was trying to create.

File: fastf1/api.py [hunk 4]
```diff
@@ -787,7 +844,7 @@ def data_in_lap(lap_n):
                 drv_data['Sector2SessionTime'][i+1]:
             drv_data['Sector2SessionTime'][i+1] = new_s2_time
         if (new_s3_time := drv_data['Time'][i] + drv_data['Sector1Time'][i+1] + drv_data['Sector2Time'][i+1] +
-                drv_data['Sector3Time'][i+1]) < drv_data['Sector3SessionTime'][i+1]:
+            drv_data['Sector3Time'][i+1]) < drv_data['Sector3SessionTime'][i+1]:  # noqa: E129,E501
```
Comment (theOehrly) [inline fastf1/api.py:847]: Lint change, should not be in here. Also, I don't think noqa's should be used for a visual change like that

File: fastf1/api.py [hunk 5]
```diff
@@ -1202,7 +1259,7 @@ def position_data(path, response=None, livedata=None):
                 .reset_index(drop=True)
             data[driver]['Status'].fillna(value='OffTrack', inplace=True)
             data[driver].loc[:, ['X', 'Y', 'Z']] = data[driver]\
-                .loc[:, ['X', 'Y', 'Z']].fillna(value=0, inplace=False)
+                        .loc[:, ['X', 'Y', 'Z']].fillna(value=0, inplace=False)
```
Comment (theOehrly) [inline fastf1/api.py:1262]: Unrelated lint change again

**Comment by theOehrly** [comment]
Sorry, I forgot that this was waiting on me, actually.

Another thing that's still missing is the updated documentation, to explain the new defaults and the new environment variable.

Also, what is your opinion on the multi-line cache info/warning message? Some people had complained that FastF1 produces too much output. Should we try and reduce this to a single-line message?

**Comment by oscr** [comment]
No problem at all. 

In my opinion maybe the default log level is a little verbose for users. When developing FF1 it's very helpful. But I can understand that a lot of it might not be interesting for a user. So maybe we should consider some work to set a default lower logging level, make some messages more compact and ensure there is an easy way to make it verbose when debugging it?

For example:
```python
>>> session = fastf1.get_session(2022, 13, 'Q')
api         WARNING 	

DEFAULT CACHE ENABLED!
	Cache directory: /home/oscaru/.cache/fastf1.
	Size: 32.0 KB
```

This per default could be either shortened or omitted depending on verbosity level.

File: docs/index.rst [hunk 6]
```diff
@@ -29,8 +29,15 @@ The module is designed around Pandas, Numpy and Matplotlib. This makes it easy
...
-information is stored locally as cached requests (be aware).
+FastF1 handles big chunks of data (~50-100mb per session). To improve performance
+data is per default cached locally as requests (be aware). The default placement
+of the cache is operating system specific:
+
+  - Windows: `%LOCALAPPDATA%\Temp\fastf1`
+  - macOS: `~/Library/Caches/fastf1`
+  - Linux: `~/.cache/fastf1` if `~/.cache` exists otherwise `~/.fastf1`
+
+You can override the default location by using `Cache.enable_cache('location')`.
```
Comment (theOehrly) [inline docs/index.rst:40]: I'd rather put this information into the docstring of the `Cache` class. If a user allready knows that `Cache` exists (or a search engine directs them there), they'll expect the information there. And we should try to not duplicate information in the docs for easier maintenance. So I'd rather phrase is something like "The default placement of the cache is operating system specific. A custom location can be set if desired. For more information see :class:\`~fastf1.api.Cache\`".
Then add the detailed explanation there, including the information about the environment variable.

**Review by theOehrly — CHANGES_REQUESTED** [review]

**Comment by theOehrly** [comment]
An easy way to set the log level has been discussed a bit in #211. The current log level is mostly very verbose to fill the long duration of loading data for the first time (can be greater than 30 seconds).
Let's just stick with the message we currently have and maybe consider a change when the whole logging part is being changed.

**Comment by oscr** [comment]
> An easy way to set the log level has been discussed a bit in #211. The current log level is mostly very verbose to fill the long duration of loading data for the first time (can be greater than 30 seconds).
> Let's just stick with the message we currently have and maybe consider a change when the whole logging part is being changed.

I would be willing to look at logging overall at a later stage if you don't feel that you really love logging and would love to do it.

**Comment by theOehrly** [comment]
OK, did you forget to mention the environment variable in the docs again, or am I just missing it when reading through it?
Apart from that, I really like how you worded the documentation update. Very concise but informative.

**Comment by oscr** [comment]
> OK, did you forget to mention the environment variable in the docs again, or am I just missing it when reading through it? Apart from that, I really like how you worded the documentation update. Very concise but informative.

No, I just totally forgot to mention that feature. I will add it.

**Comment by theOehrly** [comment]
Regarding logging, I can also look into this myself. There is actually also the need to add a certain level at which some logging is disabled in favour of raising actually raising exceptions to facilitate development.

File: fastf1/api.py [hunk 7]
```diff
@@ -97,17 +97,28 @@ class Cache:
...
-    circumstances which are usually limited to doing core developement
-    on FastF1.
+    An alternative way to set the cache directory is to configure an
+    environment variable `FASTF1_CACHE`. However, this value will be
+    ignored if `Cache.enable_cache()` is called.
+
+    If no explicit location is provied, Fast-F1 will use a default location
+    depending on operating system.
+
+        - Windows: `%LOCALAPPDATA%\\Temp\\fastf1`
```
Comment (theOehrly) [inline fastf1/api.py:115]: Sorry, there is still a minor problem that I just spotted when checking the doc build. Annoyingly enough, you need to double escape backslashes here. Therefore, you need FOUR backslashes in the docstring to render a single bachslash in html.

Comment (oscr) [reply fastf1/api.py]: Good catch! I certainly wasn't expecting `\\\\` :sweat_smile: 

Fixed!

**Review by theOehrly — APPROVED** [review]

**Comment by theOehrly** [comment]
Nice, I'll merge this now. Thank you for your work! Sorry to drag this out on minor things when the actual functionality was working fine already, but I still consider these things kind of important.

**Comment by oscr** [comment]
@theOehrly Thank you for all your help and reviews. No, thank you for taking the time and effort to make sure the details are correct. It is as you say: very important.
